### PR TITLE
Update "Data Types" Page

### LINKED ISSUE #43: Review of the docs
Here are some remarks regarding the documentation:

> Installing Noir from the Compiled Binary  

We should update the binaries

> Note: At the time of pre-alpha, Nargo supports only one backend which is Aztec-Barretenberg.  

I guess we are now in pre-beta instead of pre-alpha, and we do support another backend. However only the Aztec one is compatible with the ethereum contract.

> TornadoCash:

Section is empty, may be we want to rename it as well?

> This should generally be the default type reached for to solve problems as using a smaller integer type like u64 incurs extra range constraints and so is less efficient rather than more.

it seems convoluted to me. But being not a native english speaker, I may be totally wrong.

>Note: The Aztec backend only supports even sized integer types currently, so while using the Aztec backend, only even sized integer types such as u32, u48 will produce proofs.

This should not be true anymore, odd-sized integers should be fine.

> f a constant type that is being used in your program in changed, 

typo; 'in' should be  'IS'

> Note: This behaviour (n.b. pub in main arguments) and type will change in future releases, to catch a linearity bug in user code.

Why this behavior will change?

>An array is a data structure which allows you to group together data types

 to group together data of the same type

> Note that currently Noir only supports arrays with integer or field elements.

we do not support array of struct?

> Example: An array of Witness types cannot be grouped together with an array of Integer type

an array of Field (instead of witness)

>For loops are expressions, so each iteration of the loop produces a value which is collected into an array.
The following code **produces an array** of 10 values, each element contains the same values from 0 to 9.
let my_arr = for i in 0..10 {
    i
};

 I don't think this is working anymore

> If Expressions
    Currently, these are not supported in the language.

yes they are!

```
<	constrains one value to be less than the other			Upper bound must have a known bit size
<=	constrains one value to be less than or equal to the other	Upper bound must have a known bit size
>	constrains one value to be more than the other			Upper bound must have a known bit size
>=	constrains one value to be more than or equal to the other	Upper bound must have a known bit size
==	constrains one value to be equal to the other			Both types must not be constants
!=	constrains one value to not be equal to the other		Both types must not be constants
```

It does not constrains, it returns a bool value.
Types can be constant, 
for the known bit size, I am not sure, but it is probably the case (that it must be known)

> they compare two values. This differs from the operations such as + where the operands are used in computation.

Not sure it is correct, because you can use comparisons in computations.

> you can use the bitwise operators | and & which operate indentically for booleans, just without the short-circuiting

an example would be welcome

>Features Coming Soon
If Statement / Ternary Expression

IF statement are done.

>Currently only primitives types are available. This does not hinder expressiveness, however for large circuits, readability will be harmed without structs.

Struct are done

> Isize
Signed integers such as i32 and i64 allow one to express more circuits.

signed integers are partially supported, they should really be finalised.

### LINKED ISSUE #54: Review "Data Types" Page
- Add a "Boolean Type" section
- Move Constants & Globals sections out to the Mutability page
- Edit Tuple example to demonstrate access via destructuring more straightforwardly
- Improve reading / comprehension flow
    - Add a Private & Public Values section
    - Swap the orders to introduce Tuples first then Structs

### PR DESCRIPTION
**Pull request by Savio-Sou**
Closes https://github.com/noir-lang/docs/issues/54, Closes https://github.com/noir-lang/docs/issues/43.

# "Data Types" Page Changes
- Add a "Boolean Type" section
- Move Constants & Globals sections out to the "Mutability" page
- Simplify the Tuple example demonstrating access via destructuring
- Modify Struct examples to use less technical terms
- Improve reading / comprehension flow
    - Add a Private & Public Types section
    - Swap the orders to introduce Tuples first then Structs
    - Minor readability edits

# "Hello, World" Page Changes
- Add a reference to the "Data Types" page for more info on private & public values
- Minor fixes

### CODE REVIEW

**Comment by Globallager** [comment]
# Preview Changes
1. `git clone https://github.com/noir-lang/book.git`
2. `cd book`
3. `git fetch origin pull/42/head:pr/42`
4. `git checkout pr/42`
5. `mdbook serve --open`

**Review by vezenovm — APPROVED** [review]
These changes look good to me

File: src/language_concepts/data_types.md [hunk 1]
```diff
@@ -1,206 +1,192 @@
-# Data types
+# Data Types
 
-Noir has several varieties of data types: the primitive `Field` type, integer types, `bool`, arrays, structs, and tuple types.
+Every value in Noir is specified with a certain data type, which informs Noir tools how to work with that data when processing Noir programs.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:3]: Lets reword this. "is specified with" implies types are required to be specified for all values rather than being inferred. We can say "Every value in noir has a type which determines which operations are valid for it." The point on tooling seems unnecessary especially since the only tooling that looks at types currently is the compiler itself.

File: src/language_concepts/data_types.md [hunk 2]
```diff
@@ -1,206 +1,192 @@
...
 
-Although each value in a constraint system is fundamentally a field element, we add a layer of abstraction over this;
-each value can be _concealed_ (a private type) or _revealed_ (a public type).
+All values in Noir are fundamentally `Field` elements. For a more approachable developing experience, abstractions are added on top to introduce different data types in Noir.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:5]: "fundamentally composed of `Field` elements` would be more accurate since some values like arrays may contain multiple.

File: src/language_concepts/data_types.md [hunk 3]
```diff
@@ -1,206 +1,192 @@
...
 
-A **private value** is known only to the Prover, while, a **public value** is known by the Prover and Verifier.
-All public and private values are primitive types, though individual
-fields of a struct may also be marked private or public.
+Noir has two category of data types: primitive types (e.g. `Field`, integer, `bool`) and compound types (e.g. array, tuple, struct). Each value can either be private or public.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:7]: lets change "integer" to "integers" or "signed/unsigned integers". The compound types listed should also be plural - "arrays, tuples, and structs"

File: src/language_concepts/data_types.md [hunk 4]
```diff
@@ -1,206 +1,192 @@
...
-langauges, while the former can be thought of as a visibility modifier for the Verifier.
-So a `pub Field` would be visible to the Verifier while a `Field` would not (private is the default modifier).
+## Private & Public Types
 
-## Primitive Types
-
-Any of these types can be private (the default) or public (via a `pub` modifier).
-
-Note that private types are also referred to as witnesses.
+A **private value** is known only to the Prover, while a **public value** is known by both the Prover and Verifier. All public and private values are primitive types, though individual fields of compound types may also be marked private or public.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:11]: We don't allow marking something explicitly private since it is the default. It'd be more accurate to say fields are also private by default but can be explicitly marked public.

File: src/language_concepts/data_types.md [hunk 5]
```diff
@@ -1,206 +1,192 @@
...
-
-```rust,noplaypen
-fn main(x : Field, y : Field) {
-    let z = x + y;
-}
-```
+Public data types are treated no differently to private types apart from the fact that their values will be revealed in proofs generated. Simply changing the value of a public type will not change the circuit (where the same goes for changing values of private types as well).
 
-`x`, `y` and `z` are witness types (private `Field`). Using the `let` keyword we derived a new witness `z` which is constrained to be equal to `x + y`.
+_Private_ are also referred to as _witnesses_ sometimes.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:17]: Should be "Private values"

File: src/language_concepts/data_types.md [hunk 6]
```diff
@@ -1,206 +1,192 @@
...
 
-#### Integer Type
+> **Note:** The terms private and public when applied to a type (e.g. `pub Field`) have a different meaning than when applied to a function (e.g. `pub fn foo() {}`).
+>
+> The latter is a visibility modifier familiar to those coming from other languages, while the former can be thought of as a visibility modifier for the Verifier to interpret.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:21]: Can we re-order this so that the former is explained first? I think it'd be less confusing. If we want to introduce the more familiar function visibility first we should re-order the first sentence on line 19 then.

File: src/language_concepts/data_types.md [hunk 7]
```diff
@@ -1,206 +1,192 @@
...
-fn main(x : Field) {
-    let y = x as u24;
+fn main(x : Field, y : pub Field) -> pub Field {
+    x + y
 }
 ```
 
-`x` and `y` are both private types (witnesses), however `y` is an integer type.
-If `y` exceeds the range \\([0,2^{24}-1]\\) then any proof created will output false by the verifier.
+In this example, the first input (i.e. `x`) is private while the second input (i.e. `y`) and the return value (i.e. `x + y`) are public.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:33]: Too much use of (i.e) here. Would be more natural to remove it:
> In this example, `x` is private while `y` and the return value are public. Note that visibility is handled per-variable, so it is perfectly valid to have one input that is private and another that is public.

File: src/language_concepts/data_types.md [hunk 8]
```diff
@@ -1,206 +1,192 @@
...
 
-> **Note:** The Aztec backend supports even and odd sized integer types, so while using the Aztec backend, even sized integer types such as u32, u48, as well as odd sized integer types such as u5, u3 will produce proofs.
+> **Note:** As of this moment, public types can only be declared through parameters on `main`.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:35]: Lets remove "As of this moment." All information in the docs is as of this moment and subject to change

File: src/language_concepts/data_types.md [hunk 9]
```diff
@@ -1,206 +1,192 @@
...
+### The Field Type
 
-    // `comptime Field` can also be inferred:
-    let a = 5;
-}
-```
+The `Field` type corresponds to the native field type of the proving backend. For example, this would be a (roughly) 256-bit integer when using the default TurboPlonk backend.
 
-Note that variables declared as mutable may not be constants:
+Fields are simply defined with the keyword `Field`:
```
Comment (jfecher) [inline src/language_concepts/data_types.md:45]: Fields are not defined with the Field keyword, they are defined with `let` statements like other values. I think we can remove this sentence. Perhaps replacing it with a brief explanation of ways in which we can use fields. E.g. "Fields support integer arithmetic and are often used as the default numeric type in Noir" or similar.

File: src/language_concepts/data_types.md [hunk 10]
```diff
@@ -1,206 +1,192 @@
...
-
-    // a inferred as a private Field here
-    let mut a = 5;
+fn main(x : Field, y : Field)  {
+    let z = x + y;
 }
 ```
 
-#### Globals
+`x`, `y` and `z` are all private `Field` in this example. Using the `let` keyword we derived a new private value `z` constrained to be equal to `x + y`.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:53]: It is a bit awkward with the code quoting, but we could say "are all private `Field`s" to fix the missing plural.
"derived" -> "defined"

File: src/language_concepts/data_types.md [hunk 11]
```diff
@@ -1,206 +1,192 @@
...
 
-Noir also supports global variables. However, they must be compile-time variables. If `comptime` is not explicitly written in the type annotation the compiler will implicitly specify the declaration as compile-time. They can then be used like any other compile-time variable inside functions. The global type can also be inferred by the compiler entirely. Globals can also be used to specify array annotations for function parameters and can be imported from submodules. 
+If efficiency is of priority, `Field` should be used as the default type for solving problems. Using smaller integer types (e.g. `u64`) incurs extra range constraints.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:55]: Oh perhaps this could be used/mentioned/rearranged with the removed blurb on line 45.

File: src/language_concepts/data_types.md [hunk 12]
```diff
@@ -1,206 +1,192 @@
...
-    }
-}
-```
+> **Note:** The default TurboPlonk backend supports both even (e.g. `u16`, `u48`) and odd (e.g. `u5`, `u3`) sized integer types.
 
-#### `pub` Modifier
+### The Boolean Type
 
-A public type is a value that may change per circuit instance. Unlike Constants, changing a public value will not change the circuit.
+A boolean type in Noir has two possible values: `true` and `false`, and is specified using `bool`. For example:
```
Comment (jfecher) [inline src/language_concepts/data_types.md:75]: _The_ `bool` type in Noir has two possible values: `true` and `false`.

File: src/language_concepts/data_types.md [hunk 13]
```diff
@@ -1,206 +1,192 @@
...
 
 ```rust,noplaypen
-fn main(x : pub Field) {
-
+fn main() {
+    let t = true; // type implicitly interpreted
+    let f: bool = false; // type explicit annotated
```
Comment (jfecher) [inline src/language_concepts/data_types.md:80]: We should remove these comments. This is an odd place to start mentioning type inference when we've already used it in the examples above.

File: src/language_concepts/data_types.md [hunk 14]
```diff
@@ -1,206 +1,192 @@
...
 }
 ```
 
-As of the beta release, public types can only be declared through parameters on `main`.
-In the code, they are treated no differently to witness types.
+> **Note:** When returning a boolean value, it will show up as a value of 1 for `true` and 0 for `false` in _Verifier.toml_.
 
-> **Note:** This behaviour and type will change in future releases, to catch a linearity bug in user code.
+The main usage of the boolean type is in conditionals (e.g. `if` expressions, `constrain`). More about conditionals are covered in the [Control Flow](./control_flow.md) and [Constrain Statement](./constrain.md) sections.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:86]: This reads a bit awkwardly. Lets rephrase the first sentence to: "The boolean type is most commonly used in conditionals like `if` expressions and `constrain` statements."

File: src/language_concepts/data_types.md [hunk 15]
```diff
@@ -1,206 +1,192 @@
...
+An array is one way of grouping together values into one compound type:
 
 ```rust,noplaypen
 fn main(x : Field, y : Field) {
     let my_arr = [x, y];
 }
 ```
 
-Note that currently Noir only supports arrays with integer or `Field` elements.
+All elements in an array must be of the same type (i.e. homogeneous). That is, `Field` values cannot be grouped together with integer values in one array for example.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:102]: Since this is targetted towards beginners, lets make the example a bit less generic and use "... together with `u8` values" rather than "integer values" since that may lead to more questions for users confused with the difference between Fields and integers.

File: src/language_concepts/data_types.md [hunk 16]
```diff
@@ -1,206 +1,192 @@
...
 
-> **Example:** An array of `Field` types cannot be grouped together with an array of integer types.
+Noir currently supports arrays with `Field` or integer elements only.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:104]: I believe we also support boolean arrays (? citation needed). So perhaps this should be "arrays with primitive element types only"

File: src/language_concepts/data_types.md [hunk 17]
```diff
@@ -1,206 +1,192 @@
...
-> **Note:** This behaviour and type will change in future releases, to catch a linearity bug in user code.
+The main usage of the boolean type is in conditionals (e.g. `if` expressions, `constrain`). More about conditionals are covered in the [Control Flow](./control_flow.md) and [Constrain Statement](./constrain.md) sections.
 
 ## Compound Types
 
-Noir also supports arrays, structs, and tuples whose elements may be public or private.
+A compound type groups together multiple values into one type. Elements within a compound type can be private or public.
 
-#### Arrays
+### The Array Type
```
Comment (jfecher) [inline src/language_concepts/data_types.md:92]: This section should show what the array type looks like: `[Field; 3]` for example.

File: src/language_concepts/data_types.md [hunk 18]
```diff
@@ -1,206 +1,192 @@
...
+
+    let five = tup.0;
+    let eight = tup.3;
 }
 ```
 
-#### Tuples
+### Structs
+
+A struct is used for grouping multiple related values of different types like a tuple, with the added ability to name the pieces of data (also referred to as _fields_).
```
Comment (jfecher) [inline src/language_concepts/data_types.md:141]: This reads a bit awkwardly with the ", with" and no conjunction. Perhaps we can rephrase it:
"A struct also allows for grouping multiple values of different types. Unlike tuples, we can also name each field. Note that the usage of "field" here is referring to each element of the struct and is unrelated to the `Field` type."

So we can merge the note below into the paragraph itself. I've also removed the mention of Rust since the phrase "a struct's fields" does not originate with rust nor was it popularized by it. If we want to keep the reasoning though we could mention that this follows convention established by other programming languages.

File: src/language_concepts/data_types.md [hunk 19]
```diff
@@ -1,206 +1,192 @@
...
+
+> **Note:** The naming of _fields_ here is unrelated to Noir's `Field` type. It simply mimics Rust's convention.
+
+When defining a struct, you will name the struct and every fields within as `key: type` pairs. A struct can hold fields of different types:
```
Comment (jfecher) [inline src/language_concepts/data_types.md:145]: Defining a struct requires giving it a name and listing each field within as `key: type` pairs.


I don't think we should include "A struct can hold fields of different types" since we already mentioned that above.

File: src/language_concepts/data_types.md [hunk 20]
```diff
@@ -1,206 +1,192 @@
...
-Noir also supports tuples, which can be thought of as anonymous structs with integers as field names.
+```rust,noplaypen
+struct Animal {
+    hands: Field,
+    legs: Field,
+    eyes: u8,
+}
+```
+
+An instance of struct can then be created with concrete values in `key: value` pairs. Struct fields are accessible using their given names, instead of by their order like in arrays and tuples:
```
Comment (jfecher) [inline src/language_concepts/data_types.md:155]: We can rephrase this as:

"When initializing a struct value, its fields can be listed in any order."

File: src/language_concepts/data_types.md [hunk 21]
```diff
@@ -1,206 +1,192 @@
...
 ```rust,noplaypen
 fn main() {
-    let tuple = (1, 2);
-    let (a, b) = tuple;
+    let legs = 4;
+
+    let dog = Animal {
+        hands: 0,
+        legs,
+        eyes: 2, // type implicitly interpreted
```
Comment (jfecher) [inline src/language_concepts/data_types.md:164]: Comment can be removed

File: src/language_concepts/mutability.md [hunk 22]
```diff
@@ -43,6 +43,63 @@ fn helper(mut x: i32) {
 }
 ```
 
+### Constants
+
+A constant type is a value that does not change per circuit instance. This is different to a witness which changes per proof. If a constant type that is being used in your program is changed, then your circuit will also change.
```
Comment (jfecher) [inline src/language_concepts/mutability.md:48]: I think this whole section should be reworded, we've been avoiding the use of `const` internally since it can mean several different things. This section seems to introduce `comptime` values so I think it'd be less confusing to call them `comptime` values. We could say something along the lines of "comptime values are values that are known at compile-time" or "evaluated to a constant value at compile-time" or similar.

File: src/language_concepts/mutability.md [hunk 23]
```diff
@@ -43,6 +43,63 @@ fn helper(mut x: i32) {
...
+
+Below we show how to declare a constant value:
+
+```rust,noplaypen
+fn main() {
+    let a: comptime Field = 5;
+
+    // `comptime Field` can also be inferred:
```
Comment (jfecher) [inline src/language_concepts/mutability.md:56]: Unlike the other sections, I'm fine with this note on inference here as "comptime" is likely something that will be unfamiliar to most users and they may incorrectly assume it is not part of the type and thus cannot be inferred otherwise.

**Review by jfecher — CHANGES_REQUESTED** [review]

File: src/language_concepts/mutability.md (hunk 22 above)
Comment (Globallager) [reply src/language_concepts/mutability.md:48]: Thanks for the insights.

The "Constants" and "Globals" sections were simply cut-pasted to this page and are intended to be updated with a separate PR. I'll keep the suggestion in mind when I get to it 🙌

File: src/language_concepts/mutability.md (hunk 23 above)
Comment (Globallager) [reply src/language_concepts/mutability.md:56]: Thanks, will keep the suggestion in mind 🙌

File: src/language_concepts/data_types.md (hunk 20 above)
Comment (Globallager) [reply src/language_concepts/data_types.md:155]: Plan was to highlight accessing struct fields with names instead of e.g. Animal.0 or Animal[0], but simultaneously showcasing field listing in any order does sound like a good idea.

Revised in upcoming commit.

**Comment by Globallager** [comment]
Thanks for spending the time to review it @jfecher & @vezenovm. New commit pushed w/ screenshot attached for easy preview (open in new tab + zoom in):
Data Types

File: src/language_concepts/data_types.md [hunk 24]
```diff
@@ -30,33 +30,35 @@ fn main(x : Field, y : pub Field) -> pub Field {
...
 fn main(x : Field, y : Field)  {
     let z = x + y;
 }
 ```
 
-`x`, `y` and `z` are all private `Field` in this example. Using the `let` keyword we derived a new private value `z` constrained to be equal to `x + y`.
+`x`, `y` and `z` are all private fields in this example. Using the `let` keyword we defined a new private value `z` constrained to be equal to `x + y`.
 
-If efficiency is of priority, `Field` should be used as the default type for solving problems. Using smaller integer types (e.g. `u64`) incurs extra range constraints.
+If proving efficiency is of priority, fields should be used as a default for solving problems. Smaller integer types (e.g. `u64`) come with overflow protections at the cost of extra range constraints.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:57]: This seems like it suggests integer types will fail when they overflow, but the current behavior is to wrap around instead. It sounds like we want to look into changing this default so I'm not sure which we should mention here. Perhaps we should only mention that they incur extra range constraints.

File: src/language_concepts/data_types.md [hunk 25]
```diff
@@ -66,42 +68,54 @@ fn main(x : Field, y : u32) {
...
+The boolean type is most commonly used in conditionals like `if` expressions and `constrain` statements. More about conditionals is covered in the [Control Flow](./control_flow.md) and [Constrain Statement](./constrain.md) sections.
 
 ## Compound Types
 
 A compound type groups together multiple values into one type. Elements within a compound type can be private or public.
 
 ### The Array Type
 
-An array is one way of grouping together values into one compound type:
+An array is one way of grouping together values into one compound type. It can be implicitly inferred or explicitly defined with `[<Type>; <Size>]`:
```
Comment (jfecher) [inline src/language_concepts/data_types.md:96]: "Defined" is for defining a name to be equal to a value. Since there is no value here, it'd be more accurate to say "Array types can be be inferred or explicilty specified via the syntax `[<Type>; <Size>]`"

File: src/language_concepts/data_types.md [hunk 26]
```diff
@@ -66,42 +68,54 @@ fn main(x : Field, y : u32) {
...
 
 ```rust,noplaypen
 fn main(x : Field, y : Field) {
     let my_arr = [x, y];
+    let your_arr: [Field; 2] = [x, y];
 }
 ```
 
-All elements in an array must be of the same type (i.e. homogeneous). That is, `Field` values cannot be grouped together with integer values in one array for example.
+Here, both `my_arr` and `your_arr` are instantiated as an array of `Field` type with `2` elements.
```
Comment (jfecher) [inline src/language_concepts/data_types.md:105]: Grammar - should be "as an array containing two `Field` elements"

File: src/language_concepts/data_types.md [hunk 27]
```diff
@@ -152,16 +166,16 @@ struct Animal {
 }
 ```
 
-An instance of struct can then be created with concrete values in `key: value` pairs. Struct fields are accessible using their given names, instead of by their order like in arrays and tuples:
+An instance of struct can then be created with actual values in `<Key>: <Value>` pairs of any order. Struct fields are accessible using their given names:
```
Comment (jfecher) [inline src/language_concepts/data_types.md:169]: "An instance of a struct"
"pairs in any order"

**Review by jfecher — COMMENTED** [review]
A few more small grammar items

File: src/language_concepts/data_types.md (hunk 24 above)
Comment (Globallager) [reply src/language_concepts/data_types.md]: Thanks for the correction 🙏

**Comment by Globallager** [comment]
> A few more small grammar items

Thanks for the followup review @jfecher!

**Review by jfecher — APPROVED** [review]
Thank you!
